prov/verbs/src/verbs_ep.c

```diff
@@ -173,30 +173,32 @@ ssize_t vrb_post_send(struct vrb_ep *ep, struct ibv_send_wr *wr, uint64_t flags)
 	struct ibv_send_wr *bad_wr;
 	struct vrb_cq *cq;
 	size_t credits_to_give = 0;
+	bool dec_peer_rq_credit = false;
```

Comment:
This boolean was introduced as a result of moving the vrb_alloc_ctx() call.  If we assume that the common case is that we're usually able to send, we've added this assignment, possibly a second assignment to set it to true, and another if statement to the fast path.

The boolean has slight side effect on reading the code in the error case.  See the credit label.  There peer_rq_credits is updated based on vrb_wr_consumes_recv(), and not this variable, which throws off the code symmetry.

Basically, reverting the vrb_alloc_ctx() change should remove the need for this variable.

Reply:
I see your point but there is assumption once there is no credit there won't be anymore calls to post_send?  In a heavily loaded system, once we are out of credit and returning -EAGAIN, wouldn't the caller try again and we are in alloc ctx and free ctx error path?

Reply:
agree, we should prioritize the success pass.

Reply:
to clarify, I agree with @shefty.

Having alloc_ctx and free_ctx in the error cases should be fine. The allocation is from a pre-allocated queue, not directly from OS.

Reply:
If we return EAGAIN, the caller should try to progress the CQ first before trying to send again.  But even if the caller is spinning between reading the cq and trying to send, there's no harm in allocating and freeing the ctx.  The app is burning CPU cycles doing nothing.

Reply:
Dropping the second commit.  Need to get this fix in.

Reply:
How about adding the following line to the first commit as previously suggested by @shefty?
```
if (!ep->peer_rq_credits)
    goto freectx;
```

Reply:
anyway, this is not important and doesn't really achieve anything more than saving a flush. This needs to be revisited later. Let's keep it as it is.

Reply:
I don't completely agree.  Separating the peer_rq_credits should be part of this change.  The code flow through this function does not change peer_rq_credits -- and that's important to capture.  The reason is exactly why the bug being addressed occurred in the first place.  The wrong CQ was being read because peer_rq_credits is updated through the rx cq, not the tx cq.  But it's only updated through rxm.  I even missed this on the initial review.

The code strongly suggests that peer_rq_credits *might* be updated.  With the updated code, the right CQ is being read, but the resulting code still *looks* wrong.  It looks like the rx cq should be checked, but it's is not.  That's why peer_rq_credits should be removed entirely from the if statement.

Reply:
This will allow post send that do not require RX buffer to go through.